File: backend/src/controllers/admin.controller.ts
import type { Response } from "express";
import { AppError } from "../errors/app-error.js";
import type { AuthenticatedRequest } from "../middleware/auth.middleware.js";
import type {
  AdminCollectionValidatedRequest,
  AdminOrderUpdateValidatedRequest,
  AdminOrdersQueryRequest,
  AdminProductValidatedRequest,
  AdminProductsQueryRequest,
} from "../middleware/admin.validation.js";
import type { ErrorResponse } from "../types/product.types.js";
import type {
  AdminCollectionListResponse,
  AdminCollectionResponse,
  AdminDashboardResponse,
  AdminOrderListResponse,
  AdminOrderResponse,
  AdminProductListResponse,
  AdminProductResponse,
} from "../types/admin.types.js";
import {
  createAdminCollection,
  createAdminProduct,
  getAdminDashboard,
  listAdminCollections,
  listAdminOrders,
  listAdminProducts,
  updateAdminCollection,
  updateAdminOrder,
  updateAdminProduct,
} from "../services/admin.service.js";

function assertAdminRequest(req: AuthenticatedRequest): void {
  if (req.authUserId === undefined || req.authUserRole !== "admin") {
    throw AppError.forbidden("Admin access is required");
  }
}

export async function getDashboard(
  req: AuthenticatedRequest,
  res: Response<AdminDashboardResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);
  res.status(200).json(await getAdminDashboard());
}

export async function getProducts(
  req: AuthenticatedRequest & AdminProductsQueryRequest,
  res: Response<AdminProductListResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (req.validatedAdminProductsQuery === undefined) {
    throw AppError.badRequest("Invalid admin products query");
  }

  res.status(200).json(await listAdminProducts(req.validatedAdminProductsQuery));
}

export async function createProduct(
  req: AuthenticatedRequest & AdminProductValidatedRequest,
  res: Response<AdminProductResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (req.validatedAdminProductBody === undefined) {
    throw AppError.badRequest("Invalid product payload");
  }

  res.status(201).json(await createAdminProduct(req.validatedAdminProductBody));
}

export async function updateProduct(
  req: AuthenticatedRequest & AdminProductValidatedRequest,
  res: Response<AdminProductResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (
    req.validatedProductId === undefined ||
    req.validatedAdminProductBody === undefined
  ) {
    throw AppError.badRequest("Invalid product payload");
  }

  res
    .status(200)
    .json(await updateAdminProduct(req.validatedProductId, req.validatedAdminProductBody));
}

export async function getCollections(
  req: AuthenticatedRequest,
  res: Response<AdminCollectionListResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);
  res.status(200).json(await listAdminCollections());
}

export async function createCollection(
  req: AuthenticatedRequest & AdminCollectionValidatedRequest,
  res: Response<AdminCollectionResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (req.validatedAdminCollectionBody === undefined) {
    throw AppError.badRequest("Invalid collection payload");
  }

  res.status(201).json(await createAdminCollection(req.validatedAdminCollectionBody));
}

export async function updateCollection(
  req: AuthenticatedRequest & AdminCollectionValidatedRequest,
  res: Response<AdminCollectionResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (
    req.validatedCollectionId === undefined ||
    req.validatedAdminCollectionBody === undefined
  ) {
    throw AppError.badRequest("Invalid collection payload");
  }

  res
    .status(200)
    .json(await updateAdminCollection(req.validatedCollectionId, req.validatedAdminCollectionBody));
}

export async function getOrders(
  req: AuthenticatedRequest & AdminOrdersQueryRequest,
  res: Response<AdminOrderListResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (req.validatedAdminOrdersQuery === undefined) {
    throw AppError.badRequest("Invalid admin orders query");
  }

  res.status(200).json(await listAdminOrders(req.validatedAdminOrdersQuery));
}

export async function updateOrder(
  req: AuthenticatedRequest & AdminOrderUpdateValidatedRequest,
  res: Response<AdminOrderResponse | ErrorResponse>,
): Promise<void> {
  assertAdminRequest(req);

  if (req.validatedOrderId === undefined || req.validatedAdminOrderBody === undefined) {
    throw AppError.badRequest("Invalid order payload");
  }

  res.status(200).json(await updateAdminOrder(req.validatedOrderId, req.validatedAdminOrderBody));
}

File: backend/src/main/java/com/season/app/controller/AdminController.java
package com.season.app.controller;

import com.season.app.dto.admin.*;
import com.season.app.service.AdminService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/admin")
public class AdminController {

    @Autowired
    private AdminService adminService;

    @GetMapping("/dashboard")
    public ResponseEntity<AdminDashboardResponse> getDashboard() {
        return ResponseEntity.ok(adminService.getAdminDashboard());
    }

    @GetMapping("/products")
    public ResponseEntity<AdminProductListResponse> getProducts(
            @RequestParam(value = "page", defaultValue = "1") int page,
            @RequestParam(value = "limit", defaultValue = "20") int limit,
            @RequestParam(value = "q", required = false) String q,
            @RequestParam(value = "collectionId", required = false) String collectionId,
            @RequestParam(value = "isActive", required = false) Boolean isActive) {
            
        AdminProductsQuery query = new AdminProductsQuery();
        query.setPage(page);
        query.setLimit(limit);
        query.setQ(q);
        query.setCollectionId(collectionId);
        query.setIsActive(isActive);

        return ResponseEntity.ok(adminService.listAdminProducts(query));
    }

    @PostMapping("/products")
    public ResponseEntity<AdminProductResponse> createProduct(@Valid @RequestBody AdminProductInput input) {
        return ResponseEntity.ok(adminService.createAdminProduct(input));
    }

    @PutMapping("/products/{productId}")
    public ResponseEntity<AdminProductResponse> updateProduct(
            @PathVariable("productId") String productId,
            @Valid @RequestBody AdminProductInput input) {
        return ResponseEntity.ok(adminService.updateAdminProduct(productId, input));
    }

    @GetMapping("/collections")
    public ResponseEntity<AdminCollectionListResponse> getCollections() {
        return ResponseEntity.ok(adminService.listAdminCollections());
    }

    @PostMapping("/collections")
    public ResponseEntity<AdminCollectionResponse> createCollection(@Valid @RequestBody AdminCollectionInput input) {
        return ResponseEntity.ok(adminService.createAdminCollection(input));
    }

    @PutMapping("/collections/{collectionId}")
    public ResponseEntity<AdminCollectionResponse> updateCollection(
            @PathVariable("collectionId") String collectionId,
            @Valid @RequestBody AdminCollectionInput input) {
        return ResponseEntity.ok(adminService.updateAdminCollection(collectionId, input));
    }

    @GetMapping("/orders")
    public ResponseEntity<AdminOrderListResponse> getOrders(
            @RequestParam(value = "page", defaultValue = "1") int page,
            @RequestParam(value = "limit", defaultValue = "20") int limit,
            @RequestParam(value = "status", required = false) String status,
            @RequestParam(value = "paymentStatus", required = false) String paymentStatus) {
            
        AdminOrdersQuery query = new AdminOrdersQuery();
        query.setPage(page);
        query.setLimit(limit);
        try {
            if (status != null) query.setStatus(com.season.app.model.Order.OrderStatus.fromValue(status));
            if (paymentStatus != null) query.setPaymentStatus(com.season.app.model.Order.PaymentStatus.fromValue(paymentStatus));
        } catch (IllegalArgumentException ignored) {}

        return ResponseEntity.ok(adminService.listAdminOrders(query));
    }

    @PatchMapping("/orders/{orderId}")
    public ResponseEntity<AdminOrderResponse> updateOrder(
            @PathVariable("orderId") String orderId,
            @Valid @RequestBody AdminOrderUpdateInput input) {
        return ResponseEntity.ok(adminService.updateAdminOrder(orderId, input));
    }
}